src/core/entry.zig

```diff
@@ -142,58 +122,56 @@ pub fn verifyPoh(
 /// Based on these agave functions for conformance:
 /// - [hash_transactions](https://github.com/anza-xyz/agave/blob/161fc1965bdb4190aa2d7e36c7c745b4661b10ed/entry/src/entry.rs#L215)
 /// - [MerkleTree::new](https://github.com/anza-xyz/agave/blob/161fc1965bdb4190aa2d7e36c7c745b4661b10ed/merkle-tree/src/merkle_tree.rs#L98)
-pub fn hashTransactions(
-    allocator: std.mem.Allocator,
-    preallocated_nodes: ?*std.ArrayListUnmanaged(Hash),
-    transactions: []const Transaction,
-) Allocator.Error!Hash {
-    const LEAF_PREFIX: []const u8 = &.{0};
-    const INTERMEDIATE_PREFIX: []const u8 = &.{1};
+/// Hash a group of transactions as a merkle tree and return the root node.
+///
+/// [agave](https://github.com/anza-xyz/agave/blob/161fc1965bdb4190aa2d7e36c7c745b4661b10ed/entry/src/entry.rs#L215)
+pub fn hashTransactions(transactions: []const Transaction) Hash {
+    const LEAF_PREFIX = &sig.ledger.shred.MERKLE_HASH_PREFIX_LEAF[0];
+    const INTERMEDIATE_PREFIX = &sig.ledger.shred.MERKLE_HASH_PREFIX_NODE[0];
 
-    var num_signatures: usize = 0;
+    var num_signatures: u64 = 0;
     for (transactions) |tx| num_signatures += tx.signatures.len;
-    if (num_signatures == 0) return Hash.ZEROES;
-
-    var owned_nodes = std.ArrayListUnmanaged(Hash){};
-    defer owned_nodes.deinit(allocator);
-    const nodes = if (preallocated_nodes) |pn| pn else &owned_nodes;
-    const capacity = std.math.log2(num_signatures) + 2 * num_signatures + 1;
-    nodes.clearRetainingCapacity();
-    try nodes.ensureTotalCapacity(allocator, capacity);
+    if (num_signatures == 0) return .ZEROES;
+
+    // NOTE: The Firedancer code is currently designed around never having more
+    // than 30 transactions in a microblock, and pretty much always puts one
+    // transaction per entry. The Agave code used to batch, however has also
+    // switched to one transaction per entry recently.
+    //
+    // 1_000 txns per entry is 32kb, which is a fine amount of stack space to use.
+    // TODO: we should investigate what the true limit here is, perhaps switching
+    // to a slower but less space intensive algorithm if we're hit with strange
```

Comment:
Could try the merkle-tree stack algorithm from blake3: https://github.com/BLAKE3-team/BLAKE3/blob/master/reference_impl/reference_impl.rs#L320-L354

It requires only `node_stack: [log2(max_sigs)]Hash` memory, so given max sigs per txn is `1232 / 64 = 19`, `[25]Hash` would be enough for 1 million txns over all entries.